packages/atclient/src/atclient.c

```diff
@@ -270,14 +270,136 @@ exit: { return ret; }
+
+  cJSON *root = NULL;
+
+  if(atkey->atkeytype != ATCLIENT_ATKEY_TYPE_PUBLICKEY) {
+    ret = 1;
+    atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atkey->atkeytype != ATKEYTYPE_PUBLIC\n");
+    goto exit;
+  }
+
+  // 2. build plookup: command
+  if(bypasscache) {
+    ret = atclient_atstr_append(&cmdbuffer, "bypassCache:true:");
+    if (ret != 0) {
+      atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_atstr_append: %d\n", ret);
+      goto exit;
+    }
+  }
+
+  ret = atclient_atstr_append(&cmdbuffer, "all:");
+  if (ret != 0) {
+    atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_atstr_append: %d\n", ret);
+    goto exit;
+  }
+
+  ret = atclient_atkey_to_string(*atkey, atkeystr.str, atkeystr.len, &atkeystr.olen);
+  if (ret != 0) {
+    atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_atkey_to_string: %d\n", ret);
+    goto exit;
+  }
+
+  char *atkeystrwithoutpublic = NULL;
+  char *ptr = strstr(atkeystr.str, "public:");
+  if(ptr != NULL) {
+    atkeystrwithoutpublic = ptr + strlen("public:");
+  } else {
+    atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "Could not find \"public:\" from string \"%s\"\n", atkeystr.str);
+    goto exit;
+  }
+
+  ret = atclient_atstr_append(&cmdbuffer, atkeystrwithoutpublic);
+  if (ret != 0) {
+    atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_atstr_append: %d\n", ret);
+    goto exit;
+  }
+
+  ret = atclient_atstr_append(&cmdbuffer, "\r\n");
+  if (ret != 0) {
+    atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_atstr_append: %d\n", ret);
+    goto exit;
+  }
+
+  // 3. send plookup: command
+  ret = atclient_connection_send(&(atclient->secondary_connection), (unsigned char *)cmdbuffer.str, cmdbuffer.olen,
+                                 recv.bytes, recv.len, &recv.olen);
+  if(ret != 0) {
+    atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "atclient_connection_send: %d\n", ret);
+    goto exit;
+  }
+
+  // 4. parse response
+
+  // 4a. if recv does not start with "data:", we probably got an error
+  if (!atclient_stringutils_starts_with((char *)recv.bytes, recv.olen, "data:", 5)) {
+    ret = 1;
+    atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "recv was \"%.*s\" and did not have prefix \"data:\"\n",
+                          (int)recv.olen, recv.bytes);
+    goto exit;
+  }
+
+  char *recvwithoutdata = (char *)recv.bytes + 5;
+
+  root = cJSON_Parse(recvwithoutdata);
+  if (root == NULL) {
+    ret = 1;
+    atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "cJSON_Parse: %d\n", ret);
+    goto exit;
+  }
+
+  // 4b. set *value and *valueolen
+  cJSON *data = cJSON_GetObjectItem(root, "data");
+  if (data == NULL) {
+    ret = 1;
+    atclient_atlogger_log(TAG, ATLOGGER_LOGGING_LEVEL_ERROR, "cJSON_GetObjectItem: %d\n", ret);
+    goto exit;
+  }
+
+  memset(value, 0, valuelen);
+  strcpy(value, data->valuestring);
```

Comment:
Here we could use `memcpy`, which should be safer (or even `snprintf`). Check Git's [banned.h](https://github.com/git/git/blob/master/banned.h)

Reply:
https://github.com/atsign-foundation/at_c/pull/137/commits/9b5d0d92d319df5655c04c4e53de0eec2c0d1434